File: iid/gui/widget.py
#
# This file is part of the Infinite Improbability Drive.
#
# Copyright (C) 2009 by Jernej Kos <kostko@unimatrix-one.org>
# Copyright (C) 2009 by Anze Vavpetic <anze.vavpetic@gmail.com>
#
from OpenGL.GL import *
from OpenGL.GLUT import *
import logging

# IID imports
from iid.events import Signalizable, EventType

# Logger for this module
logger = logging.getLogger(__name__)

class Widget(Signalizable):
  """
  A paintable GUI object.
  """
  parent = None
  children = None
  
  # Widget coordinates and size
  x = 0
  y = 0
  width = 0
  height = 0
  
  # Other properties
  alpha = 1.0
  visible = True
  
  # Widget that currently has focus
  focusedChild = None
  
  def __init__(self, parent, insertChild = True):
    """
    Class constructor.
    
    @param parent: A parent widget
    """
    super(Widget, self).__init__()
    self.parent = parent
    self.children = []
    
    if parent and insertChild and isinstance(parent, Widget):
      parent.children.append(self)
    
    logger.debug("Created widget '%s' with parent '%s'." % (self.__class__.__name__, self.parent.__class__.__name__))
  
  def setPosition(self, x, y):
    """
    Changes widget's position.
    """
    self.x, self.y = x, y
    self.emit("Widget.moved", x, y)
  
  def setSize(self, width, height):
    """
    Changes widget's size.
    """
    self.width, self.height = width, height
    self.emit("Widget.resized", width, height)
  
  def containsCoordinates(self, x, y):
    """
    Returns true if this widget contains the following parent-relative
    coordinates.
    
    @param x: Parent-relative X coordinate
    @param y: Parent-relative Y coordinate
    """
    return self.visible and (self.x <= x <= self. x + self.width) and (self.y <= y <= self.y + self.height)
  
  def setFocus(self, widget = None):
    """
    Sets the focus to a widget.
    """
    if self.focusedChild is not None and self.focusedChild == widget:
      return
    
    self.parent.setFocus(self)
    
    if widget:
      self.focusedChild = widget
    
    # Emit a signal
    self.emit("Widget.focused")
  
  def eventMouseMove(self, event):
    """
    Called when a mouse event ocurrs on this widget. Should return True when
    event should be propagated to child widgets.
    
    @param event: A valid MouseMoveEvent instance
    """
    return True
  
  def eventMousePress(self, event):
    """
    Called when a mouse event ocurrs on this widget. Should return True when
    event should be propagated to child widgets.
    
    @param event: A valid MousePressEvent instance
    """
    return True
  
  def eventMouse(self, event):
    """
    Called when a mouse event ocurrs on this widget. Should return True when
    event should be propagated to child widgets.
    
    @param event: A valid MouseEvent instance
    """
    if event.eventType == EventType.MouseMove:
      return self.eventMouseMove(event)
    elif event.eventType == EventType.MousePress:
      # Widgets get focus on press events by default
      if event.state == GLUT_DOWN:
        self.setFocus()
      
      return self.eventMousePress(event)
  
  def eventKeyPressed(self, event):
    """
    Called when a key press event ocurrs on this widget. Should return True
    when event should be propagated to child widgets.
    
    @param event: A valid KeyboardEvent instance
    """
    return True
  
  def event(self, event):
    """
    Called when an event ocurrs.
    
    @param event: A valid Event instance
    """
    if event.eventType in (EventType.MouseMove, EventType.MousePress):
      event.x -= self.x
      event.y -= self.y
      
      # Do local mouse press processing
      if self.eventMouse(event):
        # Only continue when handler returns True, otherwise this event is
        # considered handled by the current widget
        for widget in self.children:
          if widget.containsCoordinates(event.x, event.y):
            widget.event(event)
            break
    elif event.eventType == EventType.Keyboard:
      # Do local key press processing first
      if self.eventKeyPressed(event) and self.focusedChild and self.focusedChild.visible:
        # Only continue when handler returns True, otherwise this event is
        # considered handled by the current widget
        self.focusedChild.event(event)
  
  def paint(self):
    """
    This method should be overridden by individual widgets to actually
    paint via OpenGL.
    """
    pass
  
  def _paint(self):
    """
    Paints ourselves and then all children.
    """
    self.paint()
    
    glPushMatrix()
    glTranslatef(self.x, self.y, 0)
    
    for child in self.children:
      if child.visible:
        child._paint()
    
    glPopMatrix()

File: iid/src/gui/widget.cpp
/*
 * This file is part of the Infinite Improbability Drive.
 *
 * Copyright (C) 2009 by Jernej Kos <kostko@unimatrix-one.org>
 * Copyright (C) 2009 by Anze Vavpetic <anze.vavpetic@gmail.com>
 */
#include "gui/widget.h"
#include "gui/painter.h"
#include "gui/manager.h"
#include "events/mouse.h"
#include "events/keyboard.h"
#include "events/dispatcher.h"

#include <boost/foreach.hpp>

namespace IID {

namespace GUI {

Widget::Widget(Widget *parent)
  : m_parent(0),
    m_position(0, 0),
    m_size(0, 0),
    m_lastMouseIn(0)
{
  reparent(parent);
}

Widget::~Widget()
{
  BOOST_FOREACH(Widget *w, m_children) {
    delete w;
  }
}

Widget *Widget::getParent() const
{
  return m_parent;
}

void Widget::reparent(Widget *parent)
{
  // Clear the old parent first
  if (m_parent) {
    m_parent->m_children.erase(std::find(m_parent->m_children.begin(), m_parent->m_children.end(), this));
    
    if (m_parent->m_lastMouseIn == this)
      m_parent->m_lastMouseIn = 0;
  }
  
  // Insert into the new parent
  if (parent) {
    parent->m_children.push_back(this);
  }
  
  m_parent = parent;
}

Vector2i Widget::mapToGlobal(int x, int y) const
{
  return mapToGlobal(Vector2i(x, y));
}

Vector2i Widget::mapToGlobal(const Vector2i &p) const
{
  if (!m_parent)
    return p;
  
  return m_parent->mapToGlobal(m_position + p);
}

void Widget::setPosition(int x, int y)
{
  m_position = Vector2i(x, y);
  m_globalPosition = mapToGlobal(0, 0);
}

Vector2i Widget::getPosition() const
{
  return m_position;
}

Vector2i Widget::getGlobalPosition() const
{
  return m_globalPosition;
}

void Widget::setSize(int width, int height)
{
  m_size = Vector2i(width, height);
  // TODO resize event
}

Vector2i Widget::getSize() const
{
  return m_size;
}

int Widget::getWidth() const
{
  return m_size[0];
}

int Widget::getHeight() const
{
  return m_size[1];
}


void Widget::setEnabled(bool enabled)
{
  m_enabled = enabled;
}

bool Widget::isEnabled() const
{
  return m_enabled;
}

Widget *Widget::childAt(int x, int y) const
{
  Vector2i p(x, y);
  
  // Search through all the children to find the right one
  BOOST_FOREACH(Widget *w, m_children) {
    if ((p.cwise() >= w->getGlobalPosition()).all() && (p.cwise() <= w->getGlobalPosition() + w->getSize()).all())
      return w;
  }
  
  return 0;
}

void Widget::paintEvent(Painter *painter)
{
  // Perform local painting first
  paint(painter);
  
  // Now propagate to children
  BOOST_FOREACH(Widget *w, m_children) {
    painter->save();
    painter->translate(w->getPosition());
    painter->setClipRegion(Vector2i(0, -1), w->getSize() + Vector2i(1, 1));
    w->paintEvent(painter);
    painter->restore();
  }
}

void Widget::paint(Painter *painter)
{
  // By default we just paint a rectangle
  painter->drawRect(0, 0, m_size[0], m_size[1]);
}

bool Widget::event(Event *event)
{
  // Dispatch event to some subwidget or take it ourselves depending on the event
  switch (event->getType()) {
    case Event::MouseEvent: {
      // Find out the widget under the mouse
      MouseEvent *ev = static_cast<MouseEvent*>(event);
      Widget *w = childAt(ev->x(), ev->y());
      if (m_lastMouseIn != w) {
        // Handle mouse enter/leave events
        if (m_lastMouseIn)
          m_lastMouseIn->leaveEvent(ev);
        
        if (w)
          w->enterEvent(ev);
        
        m_lastMouseIn = w;
      }
      
      // Check event type and propagate it accordingly
      if (w && w->event(ev))
        return true;
      
      // If not handled, handle it locally
      switch (ev->getMouseEventType()) {
        case MouseEvent::MouseMove: return mouseMoveEvent(ev);
        case MouseEvent::MouseButtonPress: return mousePressEvent(ev);
        case MouseEvent::MouseButtonRelease: return mouseReleaseEvent(ev);
      }
      break;
    }
    case Event::KeyboardEvent: {
      // TODO
      break;
    }
  }
  
  return false;
}

bool Widget::mouseMoveEvent(MouseEvent *event)
{
  return true;
}

bool Widget::mousePressEvent(MouseEvent *event)
{
  return true;
}

bool Widget::mouseReleaseEvent(MouseEvent *event)
{
  return true;
}

void Widget::enterEvent(Event *event)
{
}

void Widget::leaveEvent(Event *event)
{
}

void Widget::keyPressEvent(KeyboardEvent *event)
{
}

void Widget::keyReleaseEvent(KeyboardEvent *event)
{
}

void Widget::grabMouse()
{
  Manager::getInstance()->grabMouse(this);
}
    
void Widget::ungrabMouse()
{
  Manager::getInstance()->ungrabMouse(this);
}

}

}
